### PR TITLE
update: Rewrite the expressions router Gateway doc

### PR DESCRIPTION
**Pull request by dndx**
This is a complete rewrite of the Expressions router doc to make it easier to understand and more comprehensive.

![KAG-2288]

[KAG-2288]: https://konghq.atlassian.net/browse/KAG-2288?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

### CODE REVIEW

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 1]
```diff
@@ -0,0 +1,165 @@
...
+net.src.ip in fd00::/8
+```
+
+> **Note:** Expressions parser rejects any CIDR literal where the host portion contains
+  any non-zero bits. This means that `192.168.0.1/24` won't pass the parser check because
+  the intention of the author is unclear.
+
+### IpAddr
+
+`IpCidr` represents a single IP addresses in [IPv4 Dot-decimal notation](https://en.wikipedia.org/wiki/Dot-decimal_notation),
```
Comment (randmonkey) [inline app/_src/gateway/reference/expressions-language/language-references.md:88]:
```suggestion
`IpAddr` represents a single IP addresses in [IPv4 Dot-decimal notation](https://en.wikipedia.org/wiki/Dot-decimal_notation),
```
typo here?

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 2]
```diff
@@ -0,0 +1,165 @@
...
+| `^=`            | Prefix match          | Field value starts with constant value                                                |
+| `=^`            | Postfix match         | Field value ends with constant value                                                  |
+| `>=`            | Greater than or equal | Field value greater than or equal constant value                                      |
+| `>`             | Greater than          | Field value greater than constant value                                               |
+| `<=`            | Less than or equal    | Field value less than or equal constant value                                         |
+| `<`             | Less than             | Field value less than constant value                                                  |
+| `in`            | In                    | Field value is inside constant value                                                  |
+| `not in`        | Not in                | Field value is not inside constant value                                              |
+| `contains`      | Contains              | Field value contains constant value                                                   |
```
Comment (randmonkey) [inline app/_src/gateway/reference/expressions-language/language-references.md:138]: Could you add some explanation on the difference between `in`/`not in` and `contains`?

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 3]
```diff
@@ -0,0 +1,165 @@
...
+Here is an overview:
+
+| Operator        | Name                  | Description                                                                           |
+|-----------------|-----------------------|---------------------------------------------------------------------------------------|
+| `==`            | Equals                | Field value equals to constant value                                                  |
+| `!=`            | Not equals            | Field value does not equals to constant value                                         |
+| `~`             | Regex match           | Field value matches regex                                                             |
+| `^=`            | Prefix match          | Field value starts with constant value                                                |
+| `=^`            | Postfix match         | Field value ends with constant value                                                  |
+| `>=`            | Greater than or equal | Field value greater than or equal constant value                                      |
```
Comment (randmonkey) [inline app/_src/gateway/reference/expressions-language/language-references.md:132]:
```suggestion
| `>=`            | Greater than or equal | Field value greater than or equal to constant value                                      |
```

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 4]
```diff
@@ -0,0 +1,165 @@
...
+| `>`             | Greater than          | Field value greater than constant value                                               |
+| `<=`            | Less than or equal    | Field value less than or equal constant value                                         |
```
Comment (randmonkey) [inline app/_src/gateway/reference/expressions-language/language-references.md:134]:
```suggestion
| `<=`            | Less than or equal    | Field value less than or equal to constant value                                         |
```

File: app/_src/gateway/key-concepts/routes/expressions.md [hunk 5]
```diff
@@ -3,42 +3,37 @@ title: How to Configure Routes using Expressions
...
+Please be sure to familiar yourself with the [Expressions Language Reference](/gateway/latest/reference/expressions-language/overview/)
+before processing through the rest of this guide.
 
 ## Prerequisite
 
 Edit [kong.conf](/gateway/latest/production/kong-conf/) to contain the line `router_flavor = expressions` and restart {{site.base_gateway}}.
-Note: once you enable expressions, the match fields that traditionally exist on the Route object (such as `paths`, `methods`) will no longer
-be configurable and you must specify Expressions in the `expression` field.
+> **Note:** Once you enable expressions, the match fields that traditionally exist on the Route object (such as `paths`, `methods`) will no longer
+  be configurable and you must specify Expressions in the `expression` field. A new field `prioriry` will be made available
```
Comment (chronolaw) [inline app/_src/gateway/key-concepts/routes/expressions.md:16]:
```suggestion
  be configurable and you must specify Expressions in the `expression` field. A new field `priority` will be made available
```

File: app/_src/gateway/reference/expressions-language/language-references.md (hunk 1 above)
Comment (chronolaw) [reply app/_src/gateway/reference/expressions-language/language-references.md:88]: Should be "`IpAddr` represents a single IP address"?

File: app/_src/gateway/key-concepts/routes/expressions.md [hunk 6]
```diff
@@ -3,42 +3,37 @@ title: How to Configure Routes using Expressions
 content-type: how-to
 ---
 
-
-Expressions can describe routes or paths as patterns using logical expressions.
-This how-to guide will walk through switching to the new router, and configuring routes with the new expressive domain specific language.
-For a list of all available operators and configurable fields please review the [reference documentation](/gateway/latest/reference/router-expressions-language/).
+Configuring routes using Expressions allows for more flexibility and performance
```
[the remark this answers is not in the document]
Comment (dndx) [reply app/_src/gateway/key-concepts/routes/expressions.md:6]: I am not sure, let's let the docs team decide this one.

**Review by chronolaw — APPROVED** [review]

**Review by cloudjumpercat — APPROVED** [review]
@dndx Thanks for making all these changes! I made some copyedits and structural edits here and there if you want to read through them again. I'll approve this PR so you can merge it in if everything looks good to you, but feel free to tag me if you have any questions or anything else needs to be updated.

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 7]
```diff
@@ -0,0 +1,182 @@
...
+
+Here are the allowed combination of field types and constant types with each operator:
+
+> **Note:** Rows represents field types where columns represents constant value types.
+
+| LHS/RHS      | `String`                                | `IpCidr`       | `IpAddr` | `Int`                      | `Regex` | `Expression` |
+|--------------|-----------------------------------------|----------------|----------|----------------------------|---------|--------------|
+| `String`     | `==`, `!=`, `~`, `^=`, `=^`, `contains` | ❌              | ❌        | ❌                          | `~`     | ❌            |
+| `IpAddr`     | ❌                                       | `in`, `not in` | `==`     | ❌                          | ❌       | ❌            |
+| `Int`        | ❌                                       | ❌              | ❌        | `==`, `>=`, `>`, `<=`, `<` | ❌       | ❌            |
```
Comment (chronolaw) [inline app/_src/gateway/reference/expressions-language/language-references.md:168]: Should we have `!=`?

**Comment by dndx** [comment]
Thank you @cloudjumpercat , it was a large PR and I appreciate the help.

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 8]
```diff
@@ -0,0 +1,184 @@
...
+
+This operator is used to check the existence of a string inside another string.
+For example, `http.path contains "foo"` will return `true` if `foo` can be found anywhere inside `http.path`.
+This will match a `http.path` that looks like `/foo`, `/abc/foo`, or `/xfooy`, for example.
+
+### Type and operator semantics
+
+Here are the allowed combination of field types and constant types with each operator:
+
+> **Note:** Rows represents field types that appears on the left-hand side (LHS) of the predicate where columns represents constant value types that appears on the right-hand side (RHS) of the predicate.
```
Comment (cloudjumpercat) [inline app/_src/gateway/reference/expressions-language/language-references.md:162]:
```suggestion
> **Note:** Rows represents field types that display on the left-hand side (LHS) of the predicate where columns represents constant value types that display on the right-hand side (RHS) of the predicate.
```

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 9]
```diff
@@ -0,0 +1,184 @@
...
+| `String`                         | `==`, `!=`, `~`, `^=`, `=^`, `contains` | ❌              | ❌        | ❌                                | `~`     | ❌            |
+| `IpAddr`                         | ❌                                       | `in`, `not in` | `==`     | ❌                                | ❌       | ❌            |
+| `Int`                            | ❌                                       | ❌              | ❌        | `==`, `!=`, `>=`, `>`, `<=`, `<` | ❌       | ❌            |
+| `Expression`                     | ❌                                       | ❌              | ❌        | ❌                                | ❌       | `&&`, `      |
+
+{:.note}
+> **Notes:** 
+  * The `~` operator is described as supporting both `String ~ String` and `String ~ Regex`.
+  In reality, `Regex` constant values can only be written as `String` on the right hand side.
+  The presence of `~` operators makes the string value to be treated as a regex.
```
Comment (cloudjumpercat) [inline app/_src/gateway/reference/expressions-language/language-references.md:175]:
```suggestion
  The presence of `~` operators treats the string value as a regex.
```

File: app/_src/gateway/reference/expressions-language/language-references.md [hunk 10]
```diff
@@ -0,0 +1,184 @@
...
+  Even with the `~` operator, [`String` escape rules described above](#string) still apply and it
+  is almost always easier to use raw string literals for `~` operator as described in the [`Regex` section](#regex).
```
Comment (cloudjumpercat) [inline app/_src/gateway/reference/expressions-language/language-references.md:177]:
```suggestion
  Even with the `~` operator, [`String` escape rules described above](#string) still apply and it
  is almost always easier to use raw string literals for the `~` operator as described in the [`Regex` section](#regex).
```

**Pull request merged**